flocker/ca/test/test_ca.py

```diff
@@ -16,13 +16,296 @@
+        """
+        A ``PathError`` is raised if the certificate file path given to
+        ``UserCredential.from_path`` does not exist.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        cert_file = b"{user}.crt".format(user=self.username)
+        e = self.assertRaises(
+            PathError, UserCredential.from_path, path, self.username
+        )
+        expected = (b"Certificate file could not be opened. "
+                    b"No such file or directory "
+                    b"{path}").format(
+            path=path.child(cert_file).path)
+        self.assertEqual(str(e), expected)
+
+    def test_load_error_on_non_existent_key_file(self):
+        """
+        A ``PathError`` is raised if the key file path given to
+        ``UserCredential.from_path`` does not exist.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        cert_file = b"{user}.crt".format(user=self.username)
+        key_file = b"{user}.key".format(user=self.username)
+        crt_path = path.child(cert_file)
+        crt_file = crt_path.open(b'w')
+        crt_file.write(b"dummy")
+        crt_file.close()
+        e = self.assertRaises(
+            PathError, UserCredential.from_path, path, self.username
+        )
+        expected = (b"Private key file could not be opened. "
+                    b"No such file or directory "
+                    b"{path}").format(
+                        path=path.child(key_file).path)
+        self.assertEqual(str(e), expected)
+
+    @not_root
+    @skip_on_broken_permissions
+    def test_load_error_on_unreadable_certificate_file(self):
+        """
+        A ``PathError`` is raised if the certificate file path given to
+        ``UserCredential.from_path`` cannot be opened for reading.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        cert_file = b"{user}.crt".format(user=self.username)
+        key_file = b"{user}.key".format(user=self.username)
+        crt_path = path.child(cert_file)
+        crt_file = crt_path.open(b'w')
+        crt_file.write(b"dummy")
+        crt_file.close()
+        # make file unreadable
+        crt_path.chmod(64)
+        key_path = path.child(key_file)
+        key_file = key_path.open(b'w')
+        key_file.write(b"dummy")
+        key_file.close()
+        # make file unreadable
+        key_path.chmod(64)
+        e = self.assertRaises(
+            PathError, UserCredential.from_path, path, self.username
+        )
+        expected = (
+            b"Certificate file could not be opened. "
+            b"Permission denied {path}"
+        ).format(path=crt_path.path)
+        self.assertEqual(str(e), expected)
+
+    @not_root
+    @skip_on_broken_permissions
+    def test_load_error_on_unreadable_key_file(self):
+        """
+        A ``PathError`` is raised if the key file path given to
+        ``UserCredential.from_path`` cannot be opened for reading.
+        """
```

Comment:
This and several of the other error-case tests are basically identical.  Instead of duplicating the implementation for each credential class, it would be better to factor this code so it can be reused.  Take a look at the `make_..._tests` pattern used elsewhere in the code base and try adopting it for these tests.